php/libraries/Candidate.class.inc

```diff
@@ -281,14 +265,18 @@ class Candidate
                     );
 
         if ($config->getSetting('useExternalID') === 'true') {
-            $externalID = Candidate::_generateExternalID();
+            $externalID = (new ExternalIDGenerator())->generate();
             $setArray['ExternalID'] = $externalID;
         }
 
-        $db->insert('candidate', $setArray);
+        // Generate a new CandID and insert it into the database.
+        $candID = (new CandIDGenerator())
```

Comment:
I don't think it should be the responsibility of the ID generator to insert the whole candidate in the database. It should just return a valid, unused CandID. To circumvent the race condition, we could use LOCK tables;
```
        $db = $factory->database();
        $db->run('LOCK TABLES candidate WRITE');
        $candID = (new CandIDGenerator())->generate();
        $setArray['CandID'] = $candID;
        $db->insert('candidate', $setArray);
        $db->run('UNLOCK TABLES');
```

Reply:
Thanks, this is very helpful.

Reply:
I don't think we can (or should) lock the table. It requires permission that the default SQL account shouldn't have.

Reply:
@driusan What are your thoughts on how to approach the race condition issue?

Reply:
I don't see why it needs to be any more complicated than @PapillonMcGill's suggestion of having a try/catch inside the loop, since the CandID is a primary key.

Reply:
That's how I did it before but got a conflicting review from @xlecours. Can we land on a final approach before I redo this code again?

Reply:
where is the conflicting review from @xlecours? All I can find is him saying to delegate the loop into the generator, but there's a lot of comments so I might be missing one that's hidden somewhere..

Reply:
@driusan The conflict is from me saying to delegate the loop to the generator, because to get the exception one need to insert in the table; which should ne be done by the Generator in my opinion.

Reply:
Hmm I may be getting my thoguhts twisted on this. I'm just going to write some code and see if I understand what everybody wants.

Reply:
If that makes it impossible, I don't see the problem with keeping the loop outside of the generator. You need to be able to handle the failure case, and you can't do that without inserting, whether it's done in the generator or in the place that calls it.

Reply:
I changed it so that the try-catch insertion loop is in the Candidate class but the CandIDGenerator does an initial check on its own to make sure the ID it returns is valid in the first place.